File: frontend/src/services/api/project-members.ts
/**
 * Project Members API client — RBAC project membership management.
 * Source: specs/026-project-rbac, FR-01..FR-07
 */

import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import { apiClient } from './client';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type ProjectAssignmentAction = 'add' | 'remove';

export interface ProjectMember {
  id: string;
  projectId: string;
  userId: string;
  email: string;
  fullName: string | null;
  avatarUrl: string | null;
  isActive: boolean;
  assignedAt: string;
  assignedBy: string | null;
}

export interface ProjectMemberListResponse {
  items: ProjectMember[];
  total: number;
  nextCursor: string | null;
  hasNext: boolean;
}

export interface AddMemberRequest {
  userId: string;
}

export interface BulkAssignmentActionItem {
  projectId: string;
  action: ProjectAssignmentAction;
}

export interface BulkAssignmentRequest {
  workspaceRole?: string;
  projectAssignments: BulkAssignmentActionItem[];
}

export interface BulkAssignmentWarning {
  code: string;
  message: string;
}

export interface BulkAssignmentResponse {
  userId: string;
  workspaceRole: string | null;
  projectAssignmentsUpdated: number;
  warnings: BulkAssignmentWarning[];
}

export interface ProjectSummaryChip {
  id: string;
  name: string;
  identifier: string;
  isArchived: boolean;
}

export interface MyProjectCard {
  projectId: string;
  name: string;
  identifier: string;
  description: string | null;
  icon: string | null;
  isArchived: boolean;
  issueCount: number;
  memberCount: number;
  assignedAt: string;
}

export interface MyProjectsResponse {
  items: MyProjectCard[];
  total: number;
}

export interface ArchiveProjectRequest {
  isArchived: boolean;
}

// ---------------------------------------------------------------------------
// API client object
// ---------------------------------------------------------------------------

export const projectMembersApi = {
  /**
   * GET /workspaces/{wid}/projects/{pid}/members
   * List members of a project with optional search and pagination.
   */
  listProjectMembers(
    workspaceId: string,
    projectId: string,
    params?: { search?: string; cursor?: string; pageSize?: number }
  ): Promise<ProjectMemberListResponse> {
    return apiClient.get<ProjectMemberListResponse>(
      `/workspaces/${workspaceId}/projects/${projectId}/members`,
      { params }
    );
  },

  /**
   * POST /workspaces/{wid}/projects/{pid}/members
   * Add a workspace member to a project.
   */
  addProjectMember(
    workspaceId: string,
    projectId: string,
    payload: AddMemberRequest
  ): Promise<ProjectMember> {
    return apiClient.post<ProjectMember>(
      `/workspaces/${workspaceId}/projects/${projectId}/members`,
      payload
    );
  },

  /**
   * DELETE /workspaces/{wid}/projects/{pid}/members/{uid}
   * Remove a member from a project.
   */
  removeProjectMember(workspaceId: string, projectId: string, userId: string): Promise<void> {
    return apiClient.delete(`/workspaces/${workspaceId}/projects/${projectId}/members/${userId}`);
  },

  /**
   * PATCH /workspaces/{wid}/members/{uid}/assignments
   * Bulk update workspace role and/or project assignments for a member.
   */
  bulkUpdateAssignments(
    workspaceId: string,
    userId: string,
    payload: BulkAssignmentRequest
  ): Promise<BulkAssignmentResponse> {
    return apiClient.patch<BulkAssignmentResponse>(
      `/workspaces/${workspaceId}/members/${userId}/assignments`,
      {
        workspace_role: payload.workspaceRole,
        project_assignments: payload.projectAssignments.map((a) => ({
          project_id: a.projectId,
          action: a.action,
        })),
      }
    );
  },

  /**
   * GET /workspaces/{wid}/my-projects
   * Get projects visible to the current user (assigned or admin-all).
   */
  listMyProjects(workspaceId: string): Promise<MyProjectsResponse> {
    return apiClient.get<MyProjectsResponse>(`/workspaces/${workspaceId}/my-projects`);
  },

  /**
   * PATCH /workspaces/{wid}/members/me/last-active-project
   * Update the current user's last active project.
   */
  updateLastActiveProject(workspaceId: string, projectId: string): Promise<void> {
    return apiClient.patch(`/workspaces/${workspaceId}/members/me/last-active-project`, {
      project_id: projectId,
    });
  },

  /**
   * PATCH /workspaces/{wid}/projects/{pid}/archive
   * Archive or unarchive a project (admin/owner only).
   */
  archiveProject(
    workspaceId: string,
    projectId: string,
    payload: ArchiveProjectRequest
  ): Promise<void> {
    return apiClient.patch(`/workspaces/${workspaceId}/projects/${projectId}/archive`, {
      is_archived: payload.isArchived,
    });
  },

  /**
   * DELETE /workspaces/{wid}/members/invitations/{invId}
   * Rescind a pending invitation (admin/owner only).
   */
  rescindInvitation(workspaceId: string, invitationId: string): Promise<void> {
    return apiClient.delete(`/workspaces/${workspaceId}/members/invitations/${invitationId}`);
  },
};

// ---------------------------------------------------------------------------
// TanStack Query Keys
// ---------------------------------------------------------------------------

export const projectMemberKeys = {
  all: ['project-members'] as const,
  list: (workspaceId: string, projectId: string) =>
    [...projectMemberKeys.all, workspaceId, projectId] as const,
  myProjects: (workspaceId: string) => ['my-projects', workspaceId] as const,
};

// ---------------------------------------------------------------------------
// TanStack Query Hooks
// ---------------------------------------------------------------------------

export function useProjectMembers(
  workspaceId: string,
  projectId: string,
  params?: { search?: string; cursor?: string; pageSize?: number }
) {
  return useQuery({
    queryKey: [...projectMemberKeys.list(workspaceId, projectId), params],
    queryFn: () => projectMembersApi.listProjectMembers(workspaceId, projectId, params),
    enabled: !!workspaceId && !!projectId,
  });
}

export function useMyProjects(workspaceId: string) {
  return useQuery({
    queryKey: projectMemberKeys.myProjects(workspaceId),
    queryFn: () => projectMembersApi.listMyProjects(workspaceId),
    enabled: !!workspaceId,
  });
}

export function useAddProjectMember(workspaceId: string, projectId: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (payload: AddMemberRequest) =>
      projectMembersApi.addProjectMember(workspaceId, projectId, payload),
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: projectMemberKeys.list(workspaceId, projectId),
      });
    },
  });
}

export function useRemoveProjectMember(workspaceId: string, projectId: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (userId: string) =>
      projectMembersApi.removeProjectMember(workspaceId, projectId, userId),
    onSuccess: async () => {
      await Promise.all([
        queryClient.invalidateQueries({
          queryKey: projectMemberKeys.list(workspaceId, projectId),
        }),
        queryClient.invalidateQueries({
          queryKey: ['workspaces', workspaceId, 'members'],
        }),
      ]);
    },
  });
}

export function useBulkUpdateAssignments(workspaceId: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ userId, payload }: { userId: string; payload: BulkAssignmentRequest }) =>
      projectMembersApi.bulkUpdateAssignments(workspaceId, userId, payload),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: projectMemberKeys.all });
      queryClient.invalidateQueries({ queryKey: projectMemberKeys.myProjects(workspaceId) });
      queryClient.invalidateQueries({ queryKey: ['workspace-members', workspaceId] });
    },
  });
}

export function useRescindInvitation(workspaceId: string) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (invitationId: string) =>
      projectMembersApi.rescindInvitation(workspaceId, invitationId),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['workspace-invitations', workspaceId] });
    },
  });
}

File: backend/src/pilot_space/api/v1/routers/project_members.py
"""Project members router — US1, US4, US6.

Provides endpoints for project-scoped membership management.

Routes mounted under /workspaces/{workspace_id}/projects/{project_id}/members.
"""

from __future__ import annotations

from typing import Any
from uuid import UUID

from fastapi import APIRouter, HTTPException, Query, status

from pilot_space.api.v1.dependencies import ProjectMemberServiceDep
from pilot_space.api.v1.schemas.project_member import (
    AddProjectMemberRequest,
    ArchiveProjectRequest,
    ProjectMemberListResponse,
    ProjectMemberResponse,
    RemoveProjectMemberResponse,
)
from pilot_space.application.services.project_member import (
    AddMemberPayload,
    AlreadyProjectMemberError,
    ListMembersPayload,
    RemoveMemberPayload,
    UnauthorizedError,
)
from pilot_space.dependencies.auth import CurrentUser, SessionDep
from pilot_space.infrastructure.database.rls import set_rls_context
from pilot_space.infrastructure.logging import get_logger

logger = get_logger(__name__)

router = APIRouter()


@router.get(
    "/{workspace_id}/projects/{project_id}/members",
    response_model=ProjectMemberListResponse,
    tags=["project-members"],
)
async def list_project_members(
    workspace_id: UUID,
    project_id: UUID,
    session: SessionDep,
    current_user: CurrentUser,
    service: ProjectMemberServiceDep,
    search: str | None = Query(default=None, description="Search by name or email"),
    cursor: str | None = Query(default=None, description="Pagination cursor"),
    page_size: int = Query(default=20, ge=1, le=100),
    is_active: bool = Query(default=True, description="Filter by active status"),
) -> ProjectMemberListResponse:
    """List members assigned to a project.

    All workspace members (admin or not) may view the member list.
    """
    await set_rls_context(session, current_user.user_id, workspace_id)

    result = await service.list_members(
        ListMembersPayload(
            project_id=project_id,
            search=search,
            is_active=is_active,
            cursor=cursor,
            page_size=page_size,
        )
    )

    items = [ProjectMemberResponse.model_validate(m) for m in result.members]

    return ProjectMemberListResponse(
        items=items,
        total=result.total,
        next_cursor=result.next_cursor,
        has_next=result.has_next,
    )


@router.post(
    "/{workspace_id}/projects/{project_id}/members",
    response_model=ProjectMemberResponse,
    status_code=status.HTTP_201_CREATED,
    tags=["project-members"],
)
async def add_project_member(
    workspace_id: UUID,
    project_id: UUID,
    request: AddProjectMemberRequest,
    session: SessionDep,
    current_user: CurrentUser,
    service: ProjectMemberServiceDep,
) -> ProjectMemberResponse:
    """Add a workspace member to a project.

    Requires Admin or Owner workspace role.
    """
    await set_rls_context(session, current_user.user_id, workspace_id)

    # Resolve the requesting user's workspace role (from token claims or DB lookup)
    # We pass workspace role via a lazy lookup — the service will enforce admin gate.
    from pilot_space.dependencies.auth import get_current_session
    from pilot_space.infrastructure.database.repositories.workspace_member_repository import (
        WorkspaceMemberRepository,
    )

    ws_member_repo = WorkspaceMemberRepository(session=get_current_session())
    wm = await ws_member_repo.get_by_user_workspace(current_user.user_id, workspace_id)
    if not wm:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You are not a member of this workspace.",
        )

    try:
        member = await service.add_member(
            AddMemberPayload(
                workspace_id=workspace_id,
                project_id=project_id,
                user_id=request.user_id,
                requesting_user_id=current_user.user_id,
                requesting_user_role=wm.role.value,
            )
        )
    except UnauthorizedError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e)) from e
    except AlreadyProjectMemberError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e)) from e

    return ProjectMemberResponse(
        id=member.id,
        project_id=member.project_id,
        user_id=member.user_id,
        email=member.user.email if member.user else "",
        full_name=member.user.full_name if member.user else None,
        avatar_url=member.user.avatar_url if member.user else None,
        assigned_at=member.assigned_at,
        assigned_by=member.assigned_by,
        is_active=member.is_active,
    )


@router.delete(
    "/{workspace_id}/projects/{project_id}/members/{user_id}",
    response_model=RemoveProjectMemberResponse,
    tags=["project-members"],
)
async def remove_project_member(
    workspace_id: UUID,
    project_id: UUID,
    user_id: UUID,
    session: SessionDep,
    current_user: CurrentUser,
    service: ProjectMemberServiceDep,
) -> RemoveProjectMemberResponse:
    """Remove a member from a project (deactivate their membership).

    Requires Admin or Owner workspace role.
    """
    await set_rls_context(session, current_user.user_id, workspace_id)

    from pilot_space.dependencies.auth import get_current_session
    from pilot_space.infrastructure.database.repositories.workspace_member_repository import (
        WorkspaceMemberRepository,
    )

    ws_member_repo = WorkspaceMemberRepository(session=get_current_session())
    wm = await ws_member_repo.get_by_user_workspace(current_user.user_id, workspace_id)
    if not wm:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You are not a member of this workspace.",
        )

    try:
        removed = await service.remove_member(
            RemoveMemberPayload(
                workspace_id=workspace_id,
                project_id=project_id,
                user_id=user_id,
                requesting_user_id=current_user.user_id,
                requesting_user_role=wm.role.value,
            )
        )
    except UnauthorizedError as e:
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail=str(e)) from e

    return RemoveProjectMemberResponse(removed=removed, user_id=user_id)


@router.patch(
    "/{workspace_id}/projects/{project_id}/archive",
    tags=["project-members"],
)
async def archive_project(
    workspace_id: UUID,
    project_id: UUID,
    request: ArchiveProjectRequest,
    session: SessionDep,
    current_user: CurrentUser,
) -> dict[str, Any]:
    """Archive or unarchive a project.

    Sets is_archived and archived_at. Requires Admin or Owner.
    """
    from datetime import UTC, datetime

    from sqlalchemy import select

    from pilot_space.dependencies.auth import get_current_session
    from pilot_space.infrastructure.database.models.project import Project
    from pilot_space.infrastructure.database.models.workspace_member import WorkspaceRole
    from pilot_space.infrastructure.database.repositories.workspace_member_repository import (
        WorkspaceMemberRepository,
    )

    await set_rls_context(session, current_user.user_id, workspace_id)
    _session = get_current_session()

    ws_member_repo = WorkspaceMemberRepository(session=_session)
    wm = await ws_member_repo.get_by_user_workspace(current_user.user_id, workspace_id)
    if not wm or wm.role not in (WorkspaceRole.ADMIN, WorkspaceRole.OWNER):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only admins and owners can archive projects.",
        )

    result = await _session.execute(
        select(Project).where(
            Project.id == project_id,
            Project.workspace_id == workspace_id,
            Project.is_deleted == False,  # noqa: E712
        )
    )
    project = result.scalar_one_or_none()
    if not project:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Project not found.")

    project.is_archived = request.is_archived
    project.archived_at = datetime.now(tz=UTC) if request.is_archived else None
    await _session.flush()

    archived_at = project.archived_at
    return {
        "project_id": str(project.id),
        "is_archived": project.is_archived,
        "archived_at": archived_at.isoformat() if archived_at else None,
    }
